File: 2021/src/day03/day03.swift
import Foundation

enum Day3 {
  static func input() -> [Substring] {
    do {
      return try String(contentsOfFile: "input.txt").split(separator: "\n")
    } catch { return [] }
  }

  static func part1() -> Int {
    let data = input()
    var gamma = ""
    var epsilon = ""

    for i in 0 ... data[0].count - 1 {
      var one = 0
      var zro = 0

      for line in data {
        if line[line.index(line.startIndex, offsetBy: i)] == "1" {
          one += 1
        } else {
          zro += 1
        }
      }

      if one > zro {
        gamma += "1"
        epsilon += "0"
      } else {
        gamma += "0"
        epsilon += "1"
      }
    }

    return Int(gamma, radix: 2)! * Int(epsilon, radix: 2)!
  }

  static func part2() -> Int {
    let data = input()

    var o2 = data
    var co2 = data

    var i = 0
    while o2.count > 1 {
      let bits = o2.map { $0[$0.index($0.startIndex, offsetBy: i)] }

      var select: Character = "0"
      if bits.filter({ $0 == "1" }).count >= bits.filter({ $0 == "0" }).count {
        select = "1"
      }

      o2 = o2.filter { $0[$0.index($0.startIndex, offsetBy: i)] == select }
      i += 1
    }

    i = 0
    while co2.count > 1 {
      let bits = co2.map { $0[$0.index($0.startIndex, offsetBy: i)] }

      var select: Character = "0"
      if bits.filter({ $0 == "1" }).count < bits.filter({ $0 == "0" }).count {
        select = "1"
      }

      co2 = co2.filter { $0[$0.index($0.startIndex, offsetBy: i)] == select }
      i += 1
    }

    let o2v = Int(o2[0], radix: 2)!
    let co2v = Int(co2[0], radix: 2)!

    return o2v * co2v
  }
}

File: 2021/src/day03/day03.py
input = open("input.txt", "r").read().splitlines()


def most_common(bit_list, up_to_one=False):
    halflen = len(bit_list) / 2

    if up_to_one:
        return [
            "1" if sum([int(s[i]) for s in bit_list]) >= halflen else "0"
            for i in range(0, len(bit_list[0]))
        ]
    else:
        return [
            "1" if sum([int(s[i]) for s in bit_list]) > halflen else "0"
            for i in range(0, len(bit_list[0]))
        ]


def part1():
    common = most_common(input)
    gamma = int("".join([bit for bit in common]), 2)
    epsilon = int("".join(["0" if bit == "1" else "1" for bit in common]), 2)
    print("Part 1")
    print("  ", gamma * epsilon)


def part2():
    o2 = input.copy()
    co2 = input.copy()

    for i in range(0, len(o2[0])):
        bit = most_common(o2, up_to_one=True)[i]
        o2 = [bits for bits in o2 if bits[i] == bit]
        if len(o2) == 1:
            break

    for i in range(0, len(co2[0])):
        bit = most_common(co2, up_to_one=True)[i]
        co2 = [bits for bits in co2 if bits[i] != bit]
        if len(co2) == 1:
            break

    o2 = int(o2[0], 2)
    co2 = int(co2[0], 2)

    4125600

    print("Part 2")
    print("  ", o2 * co2)


part1()
part2()
